server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -2543,6 +2545,50 @@ public void forceMerge(final boolean flush, int maxNumSegments, boolean onlyExpu
         }
     }
 
+    @Override
+    public boolean isForceMergeOptimisticallyNoOp(int maxNumSegments, boolean onlyExpungeDeletes) throws IOException {
+        if (onlyExpungeDeletes && maxNumSegments >= 0) {
+            throw new IllegalArgumentException("only_expunge_deletes and max_num_segments are mutually exclusive");
+        }
+        final var segmentCommitInfos = SegmentInfos.readLatestCommit(indexWriter.getDirectory());
+        if (hasUncommittedChanges()) {
+            return false;
+        }
+
+        final var segmentsToMerge = new HashMap<SegmentCommitInfo, Boolean>();
+        for (int i = 0; i < segmentCommitInfos.size(); i++) {
+            final var segmentInfo = segmentCommitInfos.info(i);
+            if (onlyExpungeDeletes && segmentInfo.hasDeletions()) {
+                return false;
+            }
+            segmentsToMerge.put(segmentInfo, Boolean.TRUE);
+        }
+        if (onlyExpungeDeletes) {
+            return true;
+        }
+
+        final MergePolicy.MergeSpecification mergeSpecification;
+        if (maxNumSegments < 0) {
+            mergeSpecification = indexWriter.getConfig()
+                .getMergePolicy()
+                .findMerges(MergeTrigger.EXPLICIT, segmentCommitInfos, indexWriter);
+        } else {
+            mergeSpecification = indexWriter.getConfig()
+                .getMergePolicy()
+                .findForcedMerges(segmentCommitInfos, maxNumSegments, segmentsToMerge, indexWriter);
+        }
+        if (mergeSpecification != null && mergeSpecification.merges.isEmpty() == false) {
+            return false;
+        }
+        if (hasUncommittedChanges()) {
+            return false;
+        }
+        final var latestCommit = SegmentInfos.readLatestCommit(indexWriter.getDirectory());
+        // We check if the latest commit is the same as the one we initially read. If not, something changed while we were checking,
+        // so we return false to be safe.
+        return latestCommit.getGeneration() == segmentCommitInfos.getGeneration();
```

Comment:
Above also means that we can move this to right after the first `hasUncommittedChanges` check. So we:

1. grab commit
2. check for uncommitted, bail out if any.
3. check commit is still current, if not bail out (protecting against a commit being established between 1 and 2)

Then we know that at the specific moment we checked 2, there were no uncommitted changes and the commit at that moment was the commit we grabbed and thus we can safely check against this, knowing that a force-merge could have seen the same view and thus also do nothing if the check turns out to be a noop.